README.md
# Fortune Postgres Adapter

[![Build Status](https://img.shields.io/travis/fortunejs/fortune-postgres/master.svg?style=flat-square)](https://travis-ci.org/fortunejs/fortune-postgres)
[![npm Version](https://img.shields.io/npm/v/fortune-postgres.svg?style=flat-square)](https://www.npmjs.com/package/fortune-postgres)
[![License](https://img.shields.io/npm/l/fortune-postgres.svg?style=flat-square)](https://raw.githubusercontent.com/fortunejs/fortune-postgres/master/LICENSE)

This is a Postgres adapter for Fortune which makes use of specific Postgres functionality. Key features include:

- **Non-destructive table setup**: it will create tables and columns automatically upon connection, but will not alter columns. Data migrations are not handled by this adapter.
- **Emulates array foreign keys**: this adapter will *not* create junction tables, but instead create array columns, which is much faster than joins but lacks a database-level foreign key constraint, this is delegated to Fortune.
- **SQL query building**: it interprets arguments from Fortune's adapter interface directly, and generates optimized queries.

To use this adapter, the [database user](http://www.postgresql.org/docs/9.4/static/app-createuser.html) must be setup prior to attempting to connect, or the default `postgres` user can be used.

*This adapter, along with Fortune.js, does not implement ORM. This adapter sets up tables, and translates the adapter interface directly into SQL statements. It is a plain query builder for Postgres.*


## Requirements

- Postgres version **9.4** or newer. Older versions are untested and will not work.


## Usage

Install the `fortune-postgres` package from `npm`:

```
$ npm install fortune-postgres
```

Then use it with Fortune:

```js
const fortune = require('fortune')
const postgresAdapter = require('fortune-postgres')

const store = fortune({ ... }, {
  adapter: [
    postgresAdapter,
    {
      // options object, URL is mandatory.
      url: `postgres://${username}:${password}@${host}:${port}/${db}`
    }
  ]
})
```


## Options

- `url`: Connection URL string. **Required** if no other connection options are given. Add the query param `ssl=true` to enable SSL.
- `connection`: Connection object, see [documentation](https://github.com/brianc/node-pg-pool). This takes precendence over the URL. Optional.
- `pool`: an instance of `node-pg-pool` can be passed in directly. This takes precendence over all connection settings. Optional.
- `isNative`: Whether or not to use native bindings, requires `pg-native` module, which is an optional dependency of this one. Default: `false`.
- `typeMap`: an object keyed by type name and valued by table name.
- `primaryKeyType`: Data type of the primary key. May be `String`, `Number`, or a string for custom type. Default: `String`.
- `generatePrimaryKey`: A function that accepts one argument, the `type` of the record, and returns either a `String` or `Number`. By default,
  it returns 15 random bytes, base64 encoded in a URI-safe way. Set this to a falsey value like `null` to turn this off.
- `useForeignKeys`: Whether or not to use foreign key constraint, optional since it will only be applied to non-array fields. Default: `false`.


## Extension

The `query` field for the `options` object should be a function that accepts two arguments, the prepared SQL query and parameters, and returns an SQL query.


## SSL

SSL can be enabled by adding `ssl=true` as a query parameter the database URL (e.g. `postgres://postgres@localhost:5432/app_db?ssl=true`). If using the `connection` object, add an `ssl` property with the value `true`.


## Internal Usage

The database client pool is exposed as the `pool` property on the adapter instance, so for example, `store.adapter.pool` (or the alias `store.adapter.client`) lets you use the Postgres driver directly. This should be considered private since it is provided by the underlying implementation `node-pg-pool`, and is exposed for the sake of transparency.

For more on the API of the `pool` object, refer to [the node-pg-pool documentation](https://github.com/brianc/node-pg-pool).


## License

This software is licensed under the [MIT License](//github.com/fortunejs/fortune-postgres/blob/master/LICENSE).


CHANGELOG.md
# Changelog


##### 1.6.10 (2018-11-20)
- Fix: subtle bug in encoding buffer values on update.


##### 1.6.9 (2018-10-13)
- Fix: do not allow extraneous fields in update to cause silent error.


##### 1.6.7 (2018-09-26)
- Polish: make authorization parameters in connection URL optional.


##### 1.6.6 (2018-09-26)
- Fix: ensure that DeprecationWarning regarding Buffer constructor usage is removed.
- Polish: ensure that `client_min_messages` is set before attempting to create tables for each client.
- Polish: update dependencies.


##### 1.6.5 (2018-09-05)
- Fix: invalid format for JSON arrays.


##### 1.6.4 (2018-03-05)
- Fix: use a single client instance during a transaction.
- Bump `pg` dependency major version.


##### 1.6.3 (2017-08-14)
- Fix: inconsistent behavior when casting buffers.


##### 1.6.2 (2017-08-13)
- Polish: ignore creating non-enumerable fields if they are not denormalized inverse fields.
- Polish: allow `sslmode=require` query parameter for SSL.


##### 1.6.1 (2017-07-12)
- Fix: ignore errors that may occur if default database `postgres` does not exist.


##### 1.6.0 (2017-06-11)
- Feature: attempt to automatically create the database, so that this setup does not need to be done beforehand.


##### 1.5.1 (2017-05-10)
- Fix: alias `adapter.client` to `adapter.pool` for backwards compatibility. However, this should be considered private API.


##### 1.5.0 (2017-05-10)
- Feature: allow connection object or an instance of `node-pg-pool` to be passed in directly.


##### 1.4.8 (2017-05-10)
- Refactor: use `pool` API.


##### 1.4.6 (2016-08-23)
- Feature: proper support for custom types.


##### 1.4.4 (2016-08-23)
- Fix: check for type constructor.
- Polish: update dependencies.


##### 1.4.3 (2016-08-04)
- Polish: update dependencies.


##### 1.4.2 (2016-06-04)
- Polish: update dependencies.


##### 1.4.1 (2016-04-14)
- Fix: add default value as empty array when altering array columns.


##### 1.4.0 (2016-01-29)
- Feature: implement transactions.


##### 1.3.1 (2016-01-23)
- Feature: implement `range` option.
- Feature: implement `exists` option.
- Polish: add `not null` constraint on array fields.
- Polish: remove transpiler.


##### 1.2.4 (2015-12-31)
- Polish: make common table expression optional in create method.
- Fix: assign primary key after creating records if they are not supplied.
- Fix: improve support for numeric primary key types.
- Fix: change type of foreign keys for serial primary keys.
- Fix: omit `primaryKey` in create method if all IDs are missing.


##### 1.2.0 (2015-12-30)
- Feature: allow strings for `primaryKeyType`.
- Feature: allow `generatePrimaryKey` to be disabled.


##### 1.1.2 (2015-09-14)
- Fix: sort array fields by array length.


##### 1.1.1 (2015-09-12)
- Add parameters as second argument to `query` function.


##### 1.1.0 (2015-09-12)
- Breaking change: `query` field now accepts a function.


##### 1.0.13 (2015-09-10)
- Fix array containment query.


##### 1.0.12 (2015-09-08)
- Fix array of buffers input/output.


##### 1.0.8 (2015-08-26)
- Rename `typeMapping` to `typeMap`.


##### 1.0.7 (2015-08-24)
- Fix: `null` option bug.


##### 1.0.6 (2015-08-21)
- Feature: add missing columns on table if they don't exist yet.


##### 1.0.4 (2015-08-07)
- Bump dependency versions, use semver.
- Feature: add `typeMapping` option.


##### 1.0.2 (2015-07-17)
- Bump versions.
- Renamed package to `fortune-postgres`.


##### 1.0.1 (2015-07-02)
- Add option for foreign key constraint on non-array links. PostgreSQL itself may support foreign keys for array values in the future.


##### 1.0.0 (2015-06-29)
- Fix sort input.
- Fix update no-op.
- Moved `pg-native` as an optional dependency, no longer default.


##### 1.0.0-alpha.1 (2015-06-21)
- Initial release.


test/index.js
const testAdapter = require('fortune/test/adapter')
const adapter = require('../lib')

testAdapter(adapter, {
  url: 'postgres://postgres@localhost:5432/fortune_test',
  primaryKeyType: 'integer',
  useForeignKeys: true,
  // isNative: true,
  generatePrimaryKey: () => Math.floor(Math.random() * Math.pow(2, 16)),
  typeMap: {
    user: 'users',
    animal: 'animals'
  }
})


lib/helpers.js
'use strict'

const crypto = require('crypto')
const randomBytes = crypto.randomBytes

const primaryKeyTypes = new Set([ String, Number ])
const buffer = Buffer.from ||
  ((input, encoding) => {
    return new Buffer(input, encoding)
  })

const postgresTypeMap = new WeakMap([
  [ String, 'text' ],
  [ Number, 'double precision' ],
  [ Boolean, 'boolean' ],
  [ Date, 'timestamp' ],
  [ Object, 'jsonb' ],
  [ Buffer, 'bytea' ]
])


module.exports = {
  primaryKeyTypes, postgresTypeMap, inputValue,
  inputRecord, outputRecord, getCode
}


function inputValue (value) {
  // PostgreSQL expects a special encoding for buffers.
  if (Buffer.isBuffer(value)) return `\\x${value.toString('hex')}`

  // Array objects need to be stringified. See:
  // https://github.com/brianc/node-postgres/issues/1143
  if (Array.isArray(value)) return JSON.stringify(value)

  return value
}


function inputRecord (type, record) {
  const recordTypes = this.recordTypes
  const options = this.options
  const primaryKey = this.keys.primary
  const isArrayKey = this.keys.isArray
  const fields = recordTypes[type]
  const generatePrimaryKey = 'generatePrimaryKey' in options ?
    options.generatePrimaryKey : defaultPrimaryKey

  if (!(primaryKey in record) && generatePrimaryKey)
    record[primaryKey] = generatePrimaryKey(type)

  for (const field in fields) {
    const isArray = fields[field][isArrayKey]

    if (!(field in record)) {
      record[field] = isArray ? [] : null
      continue
    }

    record[field] = isArray ?
      record[field].map(inputValue) : inputValue(record[field])
  }

  return record
}


function outputRecord (type, record) {
  const recordTypes = this.recordTypes
  const primaryKey = this.keys.primary
  const isArrayKey = this.keys.isArray
  const typeKey = this.keys.type
  const denormalizedInverseKey = this.keys.denormalizedInverse
  const fields = recordTypes[type]
  const clone = {}

  for (const field in fields) {
    const fieldType = fields[field][typeKey]
    const fieldIsArray = fields[field][isArrayKey]
    const value = record[field]

    if (fields[field][denormalizedInverseKey]) {
      Object.defineProperty(clone, field, { value,
        writable: true, configurable: true })
      continue
    }

    if (fieldType &&
      (fieldType === Buffer || fieldType.prototype.constructor === Buffer) &&
      value && !Buffer.isBuffer(value)) {
      clone[field] = fieldIsArray ?
        value.map(outputBuffer) : outputBuffer(value)
      continue
    }

    if (field in record) clone[field] = value
  }

  clone[primaryKey] = record[primaryKey]

  return clone
}


function getCode (error) {
  return error.code || error.sqlState
}


function defaultPrimaryKey () {
  return randomBytes(15).toString('base64')
    .replace(/\+/g, '-').replace(/\//g, '_')
}


function outputBuffer (value) {
  if (Buffer.isBuffer(value)) return value
  return buffer(value.slice(2), 'hex')
}


lib/index.js
'use strict'

const pg = require('pg')
const url = require('url')
const helpers = require('./helpers')
const primaryKeyTypes = helpers.primaryKeyTypes
const postgresTypeMap = helpers.postgresTypeMap
const getCode = helpers.getCode
const inputRecord = helpers.inputRecord
const outputRecord = helpers.outputRecord
const inputValue = helpers.inputValue

// Postgres doesn't allow serial types for foreign keys, so it needs to be
// mapped to integer types.
const foreignKeyMap = {
  smallserial: 'smallint',
  serial: 'integer',
  bigserial: 'bigint'
}


/**
 * PostgreSQL adapter. In general, it's inconsistent regarding Buffer types,
 * due to the driver `pg`, so there are workarounds.
 */
module.exports = Adapter => class PostgreSQLAdapter extends Adapter {
  /**
   * Table setup happens at this stage. The default policy is completely
   * non-destructive, so tables and columns may only be added but not modified
   * in any way. Migrations are outside of the scope of this adapter.
   */
  connect () {
    const Promise = this.Promise
    const recordTypes = this.recordTypes
    const options = this.options
    const primaryKey = this.keys.primary
    const isArrayKey = this.keys.isArray
    const denormalizedInverseKey = this.keys.denormalizedInverse
    const typeKey = this.keys.type
    const linkKey = this.keys.link
    const types = Object.keys(recordTypes)

    if (!('url' in options) &&
      !('pool' in options) &&
      !('connection' in options))
      throw new Error('A connection URL is required.')

    let primaryKeyType = options.primaryKeyType || String

    if (typeof primaryKeyType !== 'string') {
      if (!primaryKeyTypes.has(primaryKeyType.prototype.constructor))
        throw new Error('The primary key type is invalid.')

      primaryKeyType = postgresTypeMap
        .get(primaryKeyType.prototype.constructor)
    }
    else primaryKeyType = primaryKeyType.toLowerCase()

    const foreignKeyType = primaryKeyType in foreignKeyMap ?
      foreignKeyMap[primaryKeyType] : primaryKeyType

    if (!('isNative' in options)) options.isNative = false
    if (!('typeMap' in options)) options.typeMap = {}

    const isNative = options.isNative
    const typeMap = options.typeMap

    const Pool = isNative ? pg.native.Pool : pg.Pool
    let connection = options.connection

    if ('pool' in options)
      this.pool = options.pool
    else {
      if (!connection) {
        const params = url.parse(options.url, true)
        const auth = params.auth ? params.auth.split(':') : []
        const user = auth[0]
        const password = auth[1]
        const database = params.pathname.split('/')[1]

        connection = {
          database,
          user,
          password,
          host: params.hostname,
          port: params.port,
          ssl: 'ssl' in params.query
        }
      }

      this.pool = new Pool(connection)
    }

    // This is overridden with a single client instance during a transaction.
    this.client = this.pool

    return (connection ?
      // Ensure that the database specified is created beforehand, if given
      // connection options.
      new Promise((resolve, reject) => {
        const defaultConnection = Object.assign({}, connection, {
          // This database is assumed to exist. If it doesn't, no problem,
          // a database creation attempt won't be made in that case.
          database: 'postgres'
        })
        const Client = isNative ? pg.native.Client : pg.Client
        const client = new Client(defaultConnection)

        client.connect(error =>
          // In case the default `postgres` database doesn't exist, just skip
          // this part.
          error ? resolve() :

          // Attempt to create the database specified in the options.
            client.query(`create database "${connection.database}"`,
            // Ignore errors and result here.
              () => {
              // Just disconnect after an attempt has been made.
                client.end((error) => error ? reject(error) : resolve())
              }))
      }) : Promise.resolve())

      // Make sure that tables exist.
      .then(() => Promise.all(types.map(type =>
        new Promise((resolve, reject) => {
          const mappedType = typeMap[type] || type
          const createTable = `create table if not exists "${mappedType}" ` +
            `("${primaryKey}" ${primaryKeyType} primary key)`

          // Noise reduction.
          this.pool.query('set client_min_messages = warning',
            error => error ? reject(error) :
            this.pool.query(createTable, error =>
              error ? reject(error) : resolve()))
        }))))

      // Get column definitions.
      .then(() => Promise.all(types.map(type =>
        new Promise((resolve, reject) => {
          const mappedType = typeMap[type] || type
          const getColumns = 'select * from information_schema.columns ' +
            `where table_name = '${mappedType}'`

          this.pool.query(getColumns, (error, result) =>
            error ? reject(error) : resolve(result))
        }))))

      // Add missing columns.
      .then(results => {
        const tableColumns = results.reduce((map, result, index) => {
          map[types[index]] = result.rows
          return map
        }, {})

        const addColumns = []
        const addColumn = (type, field) => new Promise((resolve, reject) => {
          const fieldDefinition = recordTypes[type][field]
          const isArray = fieldDefinition[isArrayKey]
          const fieldType = fieldDefinition[typeKey] &&
            fieldDefinition[typeKey].prototype.constructor
          const dataType = typeKey in fieldDefinition ?
            postgresTypeMap.get(fieldType) : foreignKeyType
          const link = fieldDefinition[linkKey]

          // Need to revisit this, when or if Postgres implements arrays of
          // foreign keys.
          const isForeignKey = link && !isArray

          const mappedType = type in typeMap ? typeMap[type] : type
          const addColumn = `alter table "${mappedType}" add column ` +
            `"${field}" ${dataType}${isArray ?
              '[] default \'{}\' not null' : ''}` +
            `${isForeignKey && options.useForeignKeys ?
              (` references "${link in typeMap ? typeMap[link] : link}" ` +
            'on delete set null') : ''}`

          this.pool.query(addColumn, error =>
            error ? reject(error) : resolve())
        })

        Object.keys(tableColumns).forEach(type => {
          Object.getOwnPropertyNames(recordTypes[type]).forEach(field => {
            // Do not include non-enumerable fields if they aren't denormalized
            // inverse fields. This may signify a hidden or input-only field.
            const descriptor = Object
              .getOwnPropertyDescriptor(recordTypes[type], field)

            if (!tableColumns[type].some(row => row.column_name === field) &&
              (descriptor.enumerable ||
                denormalizedInverseKey in recordTypes[type][field]))
              addColumns.push(addColumn(type, field))
          })
        })

        return Promise.all(addColumns)
      })

      .then(() => null)
  }


  disconnect () {
    const Promise = this.Promise

    return new Promise((resolve, reject) =>
      this.pool.end(error => error ? reject(error) : resolve()))
  }


  find (type, ids, options) {
    // Handle no-op.
    if (ids && !ids.length) return super.find()

    // Set options if falsy.
    if (!options) options = {}

    const Promise = this.Promise
    const client = this.client
    const recordTypes = this.recordTypes
    const typeMap = this.options.typeMap
    const primaryKey = this.keys.primary
    const isArrayKey = this.keys.isArray
    const fields = recordTypes[type]

    let columns = Object.keys(options.fields || {})
    columns = columns.length ?
      (columns.every(column => options.fields[column]) ?
        [ primaryKey ].concat(columns) :
        [ primaryKey ].concat(Object.keys(fields)
          .filter(field => !columns.some(column => column === field)))
      ).map(column => `"${column}"`).join(', ') : '*'

    const selectColumns = `select ${columns} from ` +
      `"${typeMap[type] || type}"`
    const query = options.query || (x => x)
    const parameters = []
    let index = 0
    let where = []
    let order = []
    let slice = ''

    if (ids) {
      where.push(`"${primaryKey}" in (${ids.map(() => {
        index++
        return `$${index}`
      }).join(', ')})`)
      Array.prototype.push.apply(parameters, ids)
    }

    for (const field in options.match) {
      const isArray = fields[field][isArrayKey]
      let value = options.match[field]

      if (!isArray) {
        if (Array.isArray(value))
          where.push(`"${field}" in (${value.map(mapValue).join(', ')})`)
        else {
          index++
          parameters.push(inputValue(value))
          where.push(`"${field}" = $${index}`)
        }
        continue
      }

      // Array containment.
      if (!Array.isArray(value)) value = [ value ]
      where.push(`"${field}" @> array[${value.map(mapValueCast).join(', ')}]`)
    }

    for (const field in options.exists) {
      const isArray = fields[field][isArrayKey]
      const value = options.exists[field]

      if (!isArray) {
        where.push(`"${field}" ${value ? 'is not null' : 'is null'}`)
        continue
      }

      where.push(`coalesce(array_length("${field}", 1), 0) ${
        value ? '> 0' : '= 0'}`)
    }

    for (const field in options.range) {
      const isArray = fields[field][isArrayKey]
      const value = options.range[field]

      if (!isArray) {
        if (value[0] != null) {
          index++
          parameters.push(value[0])
          where.push(`"${field}" >= $${index}`)
        }
        if (value[1] != null) {
          index++
          parameters.push(value[1])
          where.push(`"${field}" <= $${index}`)
        }
        continue
      }

      if (value[0] != null) {
        index++
        parameters.push(value[0])
        where.push(`coalesce(array_length("${field}", 1), 0) >= $${index}`)
      }
      if (value[1] != null) {
        index++
        parameters.push(value[1])
        where.push(`coalesce(array_length("${field}", 1), 0) <= $${index}`)
      }
    }

    where = where.length ? `where ${where.join(' and ')}` : ''

    for (const field in options.sort) {
      const isArray = fields[field][isArrayKey]
      order.push((isArray ?
        `coalesce(array_length("${field}", 1), 0) ` : `"${field}" `) +
        (options.sort[field] ? 'asc' : 'desc'))
    }

    order = order.length ? `order by ${order.join(', ')}` : ''

    if (options.limit) slice += `limit ${options.limit} `
    if (options.offset) slice += `offset ${options.offset} `

    const findRecords = query(
      `${selectColumns} ${where} ${order} ${slice}`, parameters)

    // Parallelize the find method with count method.
    return Promise.all([
      new Promise((resolve, reject) =>
        client.query(findRecords, parameters.length ? parameters : null,
          (error, result) => error ? reject(error) : resolve(result))),
      new Promise((resolve, reject) =>
        client.query(query('select count(*) from ' +
          `"${typeMap[type] || type}" ${where}`, parameters),
        parameters.length ? parameters : null,
        (error, result) => error ? reject(error) : resolve(result)))
    ])

      .then(results => {
        const records = results[0].rows.map(outputRecord.bind(this, type))
        records.count = parseInt(results[1].rows[0].count, 10)
        return records
      })

    // These functions modify the variables in this closure.

    function mapValueCast (value) {
      index++
      parameters.push(inputValue(value))

      let cast = ''

      if (Buffer.isBuffer(value))
        cast = '::bytea'
      else if (typeof value === 'number' && value % 1 === 0)
        cast = '::int'

      return `$${index}${cast}`
    }

    function mapValue (value) {
      index++
      parameters.push(inputValue(value))
      return `$${index}`
    }
  }


  create (type, records) {
    if (!records.length) return super.create()

    records = records.map(inputRecord.bind(this, type))

    const Promise = this.Promise
    const client = this.client
    const recordTypes = this.recordTypes
    const typeMap = this.options.typeMap
    const primaryKey = this.keys.primary
    const ConflictError = this.errors.ConflictError

    // Need to know if we should let Postgres handle ID generation or not.
    const hasPrimaryKey = records.every(record => primaryKey in record)

    // The sort order here doesn't really matter, as long as it's consistent.
    const orderedFields = Object.keys(recordTypes[type]).sort()

    const parameters = []
    let index = 0

    /* eslint-disable prefer-template */
    const createRecords =
      (!hasPrimaryKey ? 'with inserted as (' : '') +
      ` insert into "${typeMap[type] || type}" (` +
      (hasPrimaryKey ? [ `"${primaryKey}"` ] : [])
        .concat(orderedFields.map(field => `"${field}"`))
        .join(', ') + ') values ' +
        records.map(record => {
          if (hasPrimaryKey) parameters.push(record[primaryKey])
          Array.prototype.push.apply(parameters, orderedFields
            .map(field => record[field]))

          return `(${(hasPrimaryKey ?
            [ primaryKey ].concat(orderedFields) : orderedFields)
            .map(() => {
              index++
              return `$${index}`
            }).join(', ')})`
        }).join(', ') +
        (!hasPrimaryKey ?
          ` returning ${primaryKey}) select id from inserted` : '')
    /* eslint-enable prefer-template */

    return new Promise((resolve, reject) =>
      client.query(createRecords, parameters, (error, result) => {
        if (error) {
          const code = getCode(error)

          // Cryptic SQL error state that means unique constraint violated.
          // http://www.postgresql.org/docs/9.4/static/errcodes-appendix.html
          if (code === '23505')
            return reject(new ConflictError('Unique constraint violated.'))

          return reject(error)
        }

        if (!hasPrimaryKey)
          result.rows.forEach((result, i) => {
            records[i][primaryKey] = result[primaryKey]
          })

        return resolve(records.map(outputRecord.bind(this, type)))
      }))
  }


  update (type, updates) {
    const Promise = this.Promise
    const client = this.client
    const typeMap = this.options.typeMap
    const primaryKey = this.keys.primary
    const recordType = this.recordTypes[type]

    // This is a little bit wrong, it is only safe to update within a
    // transaction. It's not possible to put it all in one update statement,
    // since the updates may be sparse.
    return Promise.all(updates.map(update => new Promise((resolve, reject) => {
      const parameters = []
      let index = 0
      let set = []

      for (const field in update.replace) {
        if (!recordType.propertyIsEnumerable(field)) continue
        const value = update.replace[field]
        index++
        if (Array.isArray(value)) parameters.push(value.map(inputValue))
        else parameters.push(inputValue(value))
        set.push(`"${field}" = $${index}`)
      }

      for (const field in update.push) {
        if (!recordType.propertyIsEnumerable(field)) continue
        const value = update.push[field]
        index++

        if (Array.isArray(value)) {
          parameters.push(value.map(inputValue))
          set.push(`"${field}" = array_cat("${field}", $${index})`)
          continue
        }

        parameters.push(value)
        set.push(`"${field}" = array_append("${field}", $${index})`)
      }

      for (const field in update.pull) {
        if (!recordType.propertyIsEnumerable(field)) continue
        const value = update.pull[field]

        if (Array.isArray(value)) {
          // This array removal query is a modification from here:
          // http://www.depesz.com/2012/07/12/
          // waiting-for-9-3-add-array_remove-and-array_replace-functions/
          set.push(`"${field}" = array(select x from unnest("${field}") ` +
            `x where x not in (${value.map(mapValue).join(', ')}))`)
          continue
        }

        index++
        parameters.push(value)
        set.push(`"${field}" = array_remove("${field}", $${index})`)
      }

      // When no fields are to be updated, it's a no-op.
      if (!set.length) return resolve(0)

      set = `set ${set.join(', ')}`

      index++
      parameters.push(update[primaryKey])
      const updateRecord = `update "${typeMap[type] || type}" ${set} ` +
        `where "${primaryKey}" = $${index}`

      return client.query(updateRecord, parameters, (error, result) => {
        if (error) {
          const code = getCode(error)

          // If the record didn't exist, it's not an error.
          // http://www.postgresql.org/docs/9.4/static/errcodes-appendix.html
          if (code === '42703') return resolve(0)

          return reject(error)
        }

        return resolve(result.rowCount)
      })

      function mapValue (value) {
        index++
        parameters.push(value)
        return `$${index}`
      }
    })))
      .then(results => {
        return results.reduce((num, result) => {
          num += result
          return num
        }, 0)
      })
  }


  delete (type, ids) {
    if (ids && !ids.length) return super.delete()

    const Promise = this.Promise
    const client = this.client
    const typeMap = this.options.typeMap
    const primaryKey = this.keys.primary
    let index = 0

    /* eslint-disable prefer-template */
    const deleteRecords =
      `delete from "${typeMap[type] || type}"` +
      (ids ? ` where "${primaryKey}" in ` +
        `(${ids.map(() => {
          index++
          return `$${index}`
        }).join(', ')})` : '')
    /* eslint-enable prefer-template */

    return new Promise((resolve, reject) =>
      client.query(deleteRecords, ids ? ids : null,
        (error, result) => error ? reject(error) : resolve(result.rowCount)))
  }


  beginTransaction () {
    return new Promise((resolve, reject) =>
      this.pool.connect((error, client, done) => {
        if (error) return reject(error)

        const scope = Object.create(Object.getPrototypeOf(this))

        Object.assign(scope, this, {
          client,
          endTransaction (transactionError) {
            return new Promise((resolve, reject) =>
              client.query(transactionError ? 'rollback' : 'commit',
                queryError => {
                  const error = queryError || transactionError

                  if (error) {
                    done(error)
                    return reject(error)
                  }

                  // Calling done releases the client.
                  done()
                  return resolve()
                }))
          }
        })

        return client.query('begin', error =>
          error ? reject(error) : resolve(scope))
      }))
  }
}
